File: bindings/typescript/src/router.ts
import {
  create,
  fromBinary,
  toBinary,
  type DescMessage,
  type MessageShape,
} from "@bufbuild/protobuf";
import { type Any, AnySchema } from "@bufbuild/protobuf/wkt";

import { unhandled } from "./codedError";
import { Destinations } from "./destinations";
import {
  AggregateDispatch,
  ProcessManagerDispatch,
  ProjectorDispatch,
  SagaDispatch,
} from "./dispatch";
import { Ffi, type Dispatched, type Surface } from "./ffi";
import { type Session } from "./session";
import { errorResult, fromStatusBytes, type Outcome } from "./statuses";
import { CommandContext } from "./thunks";
import {
  type ApplierThunk,
  type CommandThunk,
  type PmEventThunk,
  type PmRejectionThunk,
  type ProjectorEventThunk,
  type ProjectorFinishThunk,
  type ProjectorUnknownThunk,
  type RejectionThunk,
  type SagaEventThunk,
  type SagaRejectionThunk,
} from "./thunks";
import {
  AggregateDescriptorSchema,
  CallbackEntrySchema,
  CommandContextAuxSchema,
  PmEventAuxSchema,
  PmEventEntrySchema,
  ProcessManagerDescriptorSchema,
  ProjectorDescriptorSchema,
  RejectionAuxSchema,
  RejectionEntrySchema,
  SagaDescriptorSchema,
  SagaEventAuxSchema,
} from "../gen/io/angzarr/router/ffi/v1/abi_pb";
import { BusinessResponseSchema } from "../gen/io/angzarr/v1/command_handler_pb";
import {
  type ContextualCommand,
  ContextualCommandSchema,
  type EventBook,
  EventBookSchema,
  NotificationSchema,
  type Projection,
  ProjectionSchema,
  RejectionNotificationSchema,
} from "../gen/io/angzarr/v1/types_pb";
import {
  type ProcessManagerHandleRequest,
  ProcessManagerHandleRequestSchema,
  type ProcessManagerHandleResponse,
  ProcessManagerHandleResponseSchema,
} from "../gen/io/angzarr/v1/process_manager_pb";
import {
  type SagaHandleRequest,
  SagaHandleRequestSchema,
  type SagaResponse,
  SagaResponseSchema,
} from "../gen/io/angzarr/v1/saga_pb";

// A type-erased bridge from a callback_id to a registered thunk: it receives the
// per-dispatch session and the marshaled callback inputs and returns the
// response bytes + ABI status. Throwing is the failure path — the session's
// single catch is the exception firewall.
type Invoker = (
  session: DispatchSession,
  typeUrl: string,
  payload: Uint8Array,
  aux: Uint8Array,
) => Outcome;

let abiChecked = false;

/**
 * The TypeScript binding's router: wraps the native router plus the host-side
 * callback registry the core reaches through the single callback trampoline.
 * Register a component (assigning callback ids to its thunks and handing the
 * core a serialized descriptor), then dispatch books/commands through it.
 */
export class Router {
  private readonly ptr: unknown;
  private readonly registry = new Map<number, Invoker>();
  private nextId = 0;

  constructor() {
    if (!abiChecked) {
      Ffi.init();
      abiChecked = true;
    }
    this.ptr = Ffi.routerNew();
  }

  /** The ABI version a compatible loaded cdylib reports. */
  static abiVersion(): number {
    return 1;
  }

  /** Frees the underlying native router. */
  close(): void {
    Ffi.routerFree(this.ptr);
  }

  invokerFor(callbackId: number): Invoker | undefined {
    return this.registry.get(callbackId);
  }

  private assign(invoker: Invoker): bigint {
    const id = ++this.nextId;
    this.registry.set(id, invoker);
    return BigInt(id);
  }

  // --- registration ----------------------------------------------------------

  registerAggregate<T>(d: AggregateDispatch<T>): void {
    const factory = d.rebuilder.factory;
    const desc = create(AggregateDescriptorSchema, {
      name: d.name,
      domain: d.domain,
    });
    for (const [fq, thunk] of d.rebuilder.appliers) {
      desc.appliers.push(
        create(CallbackEntrySchema, {
          fqType: fq,
          callbackId: this.assign(applierInvoker(factory, thunk)),
        }),
      );
    }
    if (d.rebuilder.snapshot) {
      desc.snapshotCallbackId = this.assign(
        applierInvoker(factory, d.rebuilder.snapshot),
      );
    }
    for (const [fq, thunk] of d.commands) {
      desc.commands.push(
        create(CallbackEntrySchema, {
          fqType: fq,
          callbackId: this.assign(commandInvoker(factory, thunk)),
        }),
      );
    }
    for (const [cmd, thunks] of d.rejections) {
      const entry = create(RejectionEntrySchema, { fqCommandType: cmd });
      for (const thunk of thunks) {
        entry.callbackIds.push(this.assign(rejectionInvoker(factory, thunk)));
      }
      desc.rejections.push(entry);
    }
    this.check(
      Ffi.register(
        "aggregate",
        this.ptr,
        toBinary(AggregateDescriptorSchema, desc),
      ),
    );
  }

  registerSaga(d: SagaDispatch): void {
    const desc = create(SagaDescriptorSchema, {
      name: d.name,
      inputDomain: d.inputDomain,
    });
    desc.targetDomains.push(...d.targets);
    for (const [fq, thunk] of d.events) {
      desc.events.push(
        create(CallbackEntrySchema, {
          fqType: fq,
          callbackId: this.assign(sagaEventInvoker(thunk)),
        }),
      );
    }
    for (const [cmd, thunks] of d.rejections) {
      const entry = create(RejectionEntrySchema, { fqCommandType: cmd });
      for (const thunk of thunks) {
        entry.callbackIds.push(this.assign(sagaRejectionInvoker(thunk)));
      }
      desc.rejections.push(entry);
    }
    this.check(
      Ffi.register("saga", this.ptr, toBinary(SagaDescriptorSchema, desc)),
    );
  }

  registerProjector<T>(d: ProjectorDispatch<T>): void {
    const factory = d.factory;
    const desc = create(ProjectorDescriptorSchema, { name: d.name });
    desc.domains.push(...d.domains);
    for (const [fq, thunk] of d.events) {
      desc.events.push(
        create(CallbackEntrySchema, {
          fqType: fq,
          callbackId: this.assign(projectorEventInvoker(factory, thunk)),
        }),
      );
    }
    if (d.unknown) {
      desc.unknownCallbackId = this.assign(projectorUnknownInvoker(d.unknown));
    }
    if (d.finisher) {
      desc.finishCallbackId = this.assign(
        projectorFinishInvoker(factory, d.finisher),
      );
    }
    this.check(
      Ffi.register(
        "projector",
        this.ptr,
        toBinary(ProjectorDescriptorSchema, desc),
      ),
    );
  }

  registerProcessManager<T>(d: ProcessManagerDispatch<T>): void {
    const factory = d.rebuilder.factory;
    const desc = create(ProcessManagerDescriptorSchema, {
      name: d.name,
      pmDomain: d.pmDomain,
    });
    for (const [fq, thunk] of d.rebuilder.appliers) {
      desc.appliers.push(
        create(CallbackEntrySchema, {
          fqType: fq,
          callbackId: this.assign(applierInvoker(factory, thunk)),
        }),
      );
    }
    if (d.rebuilder.snapshot) {
      desc.snapshotCallbackId = this.assign(
        applierInvoker(factory, d.rebuilder.snapshot),
      );
    }
    for (const [sourceDomain, byType] of d.handlers) {
      for (const [fq, thunk] of byType) {
        desc.events.push(
          create(PmEventEntrySchema, {
            inputDomain: sourceDomain,
            fqType: fq,
            callbackId: this.assign(pmEventInvoker(factory, thunk)),
          }),
        );
      }
    }
    for (const [cmd, thunks] of d.rejections) {
      const entry = create(RejectionEntrySchema, { fqCommandType: cmd });
      for (const thunk of thunks) {
        entry.callbackIds.push(this.assign(pmRejectionInvoker(factory, thunk)));
      }
      desc.rejections.push(entry);
    }
    this.check(
      Ffi.register(
        "processManager",
        this.ptr,
        toBinary(ProcessManagerDescriptorSchema, desc),
      ),
    );
  }

  private check(ret: number): void {
    if (ret !== 0) {
      throw fromStatusBytes(null, ret);
    }
  }

  // --- dispatch --------------------------------------------------------------

  dispatch(
    command: ContextualCommand,
  ): MessageShape<typeof BusinessResponseSchema> {
    return this.parse(
      this.dispatchVia("aggregate", toBinary(ContextualCommandSchema, command)),
      BusinessResponseSchema,
      "BusinessResponse",
    );
  }

  dispatchSaga(request: SagaHandleRequest): SagaResponse {
    return this.parse(
      this.dispatchVia("saga", toBinary(SagaHandleRequestSchema, request)),
      SagaResponseSchema,
      "SagaResponse",
    );
  }

  dispatchProjector(book: EventBook): Projection {
    return this.parse(
      this.dispatchVia("projector", toBinary(EventBookSchema, book)),
      ProjectionSchema,
      "Projection",
    );
  }

  dispatchProcessManager(
    request: ProcessManagerHandleRequest,
  ): ProcessManagerHandleResponse {
    return this.parse(
      this.dispatchVia(
        "processManager",
        toBinary(ProcessManagerHandleRequestSchema, request),
      ),
      ProcessManagerHandleResponseSchema,
      "ProcessManagerHandleResponse",
    );
  }

  private dispatchVia(surface: Surface, request: Uint8Array): Dispatched {
    return Ffi.dispatch(surface, this.ptr, new DispatchSession(this), request);
  }

  private parse<Desc extends DescMessage>(
    d: Dispatched,
    schema: Desc,
    what: string,
  ): MessageShape<Desc> {
    if (d.status !== 0) {
      throw fromStatusBytes(d.response, d.status);
    }
    try {
      return fromBinary(schema, d.response);
    } catch (e) {
      throw unhandled(
        `unmarshal ${what}: ${e instanceof Error ? e.message : String(e)}`,
      );
    }
  }
}

// --- session ----------------------------------------------------------------

/**
 * One dispatch's host-side state object, reached from callbacks via host_ctx.
 * The rebuilt state is created lazily by the first stateful callback (all
 * callbacks in one dispatch share it).
 */
class DispatchSession implements Session {
  private state: unknown;
  private hasState = false;

  constructor(readonly router: Router) {}

  ensureState<T>(factory: () => T): T {
    if (!this.hasState) {
      this.state = factory();
      this.hasState = true;
    }
    return this.state as T;
  }

  handleCallback(
    callbackId: number,
    typeUrl: string,
    payload: Uint8Array,
    aux: Uint8Array,
  ): Outcome {
    const invoker = this.router.invokerFor(callbackId);
    if (!invoker) {
      return errorResult(
        unhandled(`no host callback registered for id ${callbackId}`),
      );
    }
    try {
      return invoker(this, typeUrl, payload, aux);
    } catch (e) {
      return errorResult(e);
    }
  }
}

// --- invokers (type-erased bridges; thunks stay typed) -----------------------

function anyOf(typeUrl: string, payload: Uint8Array): Any {
  return create(AnySchema, { typeUrl, value: payload });
}

const OK: Outcome = { response: null, status: Ffi.STATUS_OK };
const OK_EMPTY: Outcome = { response: null, status: Ffi.STATUS_OK_EMPTY };

function applierInvoker<T>(factory: () => T, thunk: ApplierThunk<T>): Invoker {
  return (session, typeUrl, payload) => {
    thunk(session.ensureState(factory), anyOf(typeUrl, payload));
    return OK;
  };
}

function commandInvoker<T>(factory: () => T, thunk: CommandThunk<T>): Invoker {
  return (session, typeUrl, payload, aux) => {
    const cax = fromBinary(CommandContextAuxSchema, aux);
    const cctx = new CommandContext(cax.nextSequence, cax.hadPriorEvents);
    const book = thunk(
      anyOf(typeUrl, payload),
      session.ensureState(factory),
      cctx,
    );
    return book === undefined
      ? OK_EMPTY
      : { response: toBinary(EventBookSchema, book), status: Ffi.STATUS_OK };
  };
}

function rejectionInvoker<T>(
  factory: () => T,
  thunk: RejectionThunk<T>,
): Invoker {
  return (session, _typeUrl, _payload, aux) => {
    const rax = fromBinary(RejectionAuxSchema, aux);
    const n = fromBinary(NotificationSchema, rax.notification);
    const rej = fromBinary(RejectionNotificationSchema, rax.rejection);
    const cctx = new CommandContext(
      rax.cctx?.nextSequence ?? 0,
      rax.cctx?.hadPriorEvents ?? false,
    );
    const resp = thunk(n, rej, session.ensureState(factory), cctx);
    return resp === undefined
      ? OK_EMPTY
      : {
          response: toBinary(BusinessResponseSchema, resp),
          status: Ffi.STATUS_OK,
        };
  };
}

function projectorEventInvoker<T>(
  factory: () => T,
  thunk: ProjectorEventThunk<T>,
): Invoker {
  return (session, typeUrl, payload) => {
    thunk(session.ensureState(factory), anyOf(typeUrl, payload));
    return OK;
  };
}

function projectorFinishInvoker<T>(
  factory: () => T,
  thunk: ProjectorFinishThunk<T>,
): Invoker {
  return (session, _typeUrl, payload) => {
    const book =
      payload.length > 0
        ? fromBinary(EventBookSchema, payload)
        : create(EventBookSchema);
    const proj = thunk(session.ensureState(factory), book);
    return {
      response: toBinary(ProjectionSchema, proj),
      status: Ffi.STATUS_OK,
    };
  };
}

function projectorUnknownInvoker(thunk: ProjectorUnknownThunk): Invoker {
  return (_session, typeUrl) => {
    thunk(typeUrl);
    return OK;
  };
}

function sagaEventInvoker(thunk: SagaEventThunk): Invoker {
  return (_session, typeUrl, payload, aux) => {
    const sax = fromBinary(SagaEventAuxSchema, aux);
    const dests = new Destinations(sax.destinationSequences);
    const emission = thunk(anyOf(typeUrl, payload), dests);
    const resp = create(SagaResponseSchema, {
      commands: emission.commands,
      events: emission.events,
    });
    return {
      response: toBinary(SagaResponseSchema, resp),
      status: Ffi.STATUS_OK,
    };
  };
}

function sagaRejectionInvoker(thunk: SagaRejectionThunk): Invoker {
  return (_session, _typeUrl, _payload, aux) => {
    const rax = fromBinary(RejectionAuxSchema, aux);
    const n = fromBinary(NotificationSchema, rax.notification);
    const rej = fromBinary(RejectionNotificationSchema, rax.rejection);
    const resp = create(SagaResponseSchema, { events: thunk(n, rej) });
    return {
      response: toBinary(SagaResponseSchema, resp),
      status: Ffi.STATUS_OK,
    };
  };
}

function pmEventInvoker<T>(factory: () => T, thunk: PmEventThunk<T>): Invoker {
  return (session, typeUrl, payload, aux) => {
    const pax = fromBinary(PmEventAuxSchema, aux);
    const dests = new Destinations(pax.destinationSequences);
    const resp = thunk(
      anyOf(typeUrl, payload),
      session.ensureState(factory),
      dests,
    );
    return {
      response: toBinary(ProcessManagerHandleResponseSchema, resp),
      status: Ffi.STATUS_OK,
    };
  };
}

function pmRejectionInvoker<T>(
  factory: () => T,
  thunk: PmRejectionThunk<T>,
): Invoker {
  return (session, _typeUrl, _payload, aux) => {
    const rax = fromBinary(RejectionAuxSchema, aux);
    const n = fromBinary(NotificationSchema, rax.notification);
    const rej = fromBinary(RejectionNotificationSchema, rax.rejection);
    const r = thunk(n, rej, session.ensureState(factory));
    const resp = create(ProcessManagerHandleResponseSchema, {
      processEvents: r.processEvents,
    });
    if (r.escalation) {
      resp.notification = r.escalation;
    }
    return {
      response: toBinary(ProcessManagerHandleResponseSchema, resp),
      status: Ffi.STATUS_OK,
    };
  };
}

File: bindings/csharp/src/Router.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Runtime.InteropServices;
using System.Threading;
using Google.Protobuf;
using Google.Protobuf.WellKnownTypes;
using Abi = Io.Angzarr.Router.Ffi.V1;

namespace Angzarr.Router;

/// <summary>
/// The C# binding's router: wraps the native router plus the host-side callback
/// registry the core reaches through the single callback gateway. Register a
/// component (assigning callback ids to its thunks and handing the core a
/// serialized descriptor), then dispatch books/commands through it.
///
/// <para>The dispatch surfaces are generic in the component's state message, so
/// the generated wiring and handler thunks are statically typed. The one
/// unavoidable erasing cast — the FFI registry is keyed by an opaque
/// callback_id, not a type — lives in the invoker adapters below
/// (<c>(TState)session.EnsureState(...)</c>), guaranteed correct because the same
/// factory produces the state.</para>
/// </summary>
public sealed class Router : IDisposable
{
    private readonly IntPtr _ptr;
    private readonly ConcurrentDictionary<ulong, Invoker> _registry = new();
    private readonly object _lock = new();
    private long _nextId;

    public Router() => _ptr = Ffi.RouterNew();

    /// <summary>The ABI version the loaded cdylib reports (always 1 for a
    /// compatible lib).</summary>
    public static int AbiVersion() => 1;

    public void Dispose() => Ffi.RouterFree(_ptr);

    internal Invoker? InvokerFor(ulong callbackId) =>
        _registry.TryGetValue(callbackId, out var inv) ? inv : null;

    private ulong Assign(Invoker invoker)
    {
        var id = (ulong)Interlocked.Increment(ref _nextId);
        _registry[id] = invoker;
        return id;
    }

    // --- registration --------------------------------------------------------

    public void RegisterAggregate<TState>(AggregateDispatch<TState> d)
        where TState : class, IMessage
    {
        lock (_lock)
        {
            var factory = d.Rebuilder.Factory;
            var desc = new Abi.AggregateDescriptor { Name = d.Name, Domain = d.Domain };
            foreach (var (key, thunk) in d.Rebuilder.Appliers)
            {
                desc.Appliers.Add(CallbackEntry(key, Assign(ApplierInvoker(factory, thunk))));
            }
            if (d.Rebuilder.Snapshot != null)
            {
                desc.SnapshotCallbackId = Assign(ApplierInvoker(factory, d.Rebuilder.Snapshot));
            }
            foreach (var (key, thunk) in d.Commands)
            {
                desc.Commands.Add(CallbackEntry(key, Assign(CommandInvoker(factory, thunk))));
            }
            foreach (var (cmd, thunks) in d.Rejections)
            {
                var entry = new Abi.RejectionEntry { FqCommandType = cmd };
                foreach (var thunk in thunks)
                {
                    entry.CallbackIds.Add(Assign(RejectionInvoker(factory, thunk)));
                }
                desc.Rejections.Add(entry);
            }
            Check(Ffi.RegisterAggregate(_ptr, desc.ToByteArray()));
        }
    }

    public void RegisterProjector<TState>(ProjectorDispatch<TState> d)
        where TState : class, IMessage
    {
        lock (_lock)
        {
            var factory = d.Factory;
            var desc = new Abi.ProjectorDescriptor { Name = d.Name };
            desc.Domains.AddRange(d.Domains);
            foreach (var (key, thunk) in d.Events)
            {
                desc.Events.Add(CallbackEntry(key, Assign(ProjectorEventInvoker(factory, thunk))));
            }
            if (d.Unknown != null)
            {
                desc.UnknownCallbackId = Assign(ProjectorUnknownInvoker(d.Unknown));
            }
            if (d.FinishThunk != null)
            {
                desc.FinishCallbackId = Assign(ProjectorFinishInvoker(factory, d.FinishThunk));
            }
            Check(Ffi.RegisterProjector(_ptr, desc.ToByteArray()));
        }
    }

    public void RegisterSaga(SagaDispatch d)
    {
        lock (_lock)
        {
            var desc = new Abi.SagaDescriptor { Name = d.Name, InputDomain = d.InputDomain };
            desc.TargetDomains.AddRange(d.Targets);
            foreach (var (key, thunk) in d.Events)
            {
                desc.Events.Add(CallbackEntry(key, Assign(SagaEventInvoker(thunk))));
            }
            foreach (var (cmd, thunks) in d.Rejections)
            {
                var entry = new Abi.RejectionEntry { FqCommandType = cmd };
                foreach (var thunk in thunks)
                {
                    entry.CallbackIds.Add(Assign(SagaRejectionInvoker(thunk)));
                }
                desc.Rejections.Add(entry);
            }
            Check(Ffi.RegisterSaga(_ptr, desc.ToByteArray()));
        }
    }

    public void RegisterProcessManager<TState>(ProcessManagerDispatch<TState> d)
        where TState : class, IMessage
    {
        lock (_lock)
        {
            var factory = d.Rebuilder.Factory;
            var desc = new Abi.ProcessManagerDescriptor { Name = d.Name, PmDomain = d.PmDomain };
            foreach (var (key, thunk) in d.Rebuilder.Appliers)
            {
                desc.Appliers.Add(CallbackEntry(key, Assign(ApplierInvoker(factory, thunk))));
            }
            if (d.Rebuilder.Snapshot != null)
            {
                desc.SnapshotCallbackId = Assign(ApplierInvoker(factory, d.Rebuilder.Snapshot));
            }
            foreach (var (sourceDomain, byType) in d.Handlers)
            {
                foreach (var (fqType, thunk) in byType)
                {
                    desc.Events.Add(
                        new Abi.PmEventEntry
                        {
                            InputDomain = sourceDomain,
                            FqType = fqType,
                            CallbackId = Assign(PmEventInvoker(factory, thunk)),
                        }
                    );
                }
            }
            foreach (var (cmd, thunks) in d.Rejections)
            {
                var entry = new Abi.RejectionEntry { FqCommandType = cmd };
                foreach (var thunk in thunks)
                {
                    entry.CallbackIds.Add(Assign(PmRejectionInvoker(factory, thunk)));
                }
                desc.Rejections.Add(entry);
            }
            Check(Ffi.RegisterProcessManager(_ptr, desc.ToByteArray()));
        }
    }

    private static Abi.CallbackEntry CallbackEntry(string fqType, ulong id) =>
        new() { FqType = fqType, CallbackId = id };

    private static void Check(int ret)
    {
        if (ret != 0)
        {
            throw Statuses.FromStatusBytes(null, ret);
        }
    }

    // --- dispatch ------------------------------------------------------------

    public BusinessResponse Dispatch(ContextualCommand command) =>
        Parse(DispatchVia(command, Ffi.Dispatch), BusinessResponse.Parser, "BusinessResponse");

    public SagaResponse DispatchSaga(SagaHandleRequest request) =>
        Parse(DispatchVia(request, Ffi.DispatchSaga), SagaResponse.Parser, "SagaResponse");

    public Projection DispatchProjector(EventBook book) =>
        Parse(DispatchVia(book, Ffi.DispatchProjector), Projection.Parser, "Projection");

    public ProcessManagerHandleResponse DispatchProcessManager(
        ProcessManagerHandleRequest request
    ) =>
        Parse(
            DispatchVia(request, Ffi.DispatchProcessManager),
            ProcessManagerHandleResponse.Parser,
            "ProcessManagerHandleResponse"
        );

    private Ffi.Dispatched DispatchVia(
        IMessage request,
        Func<IntPtr, IntPtr, byte[], Ffi.Dispatched> call
    )
    {
        var handle = GCHandle.Alloc(new Session(this));
        try
        {
            return call(_ptr, GCHandle.ToIntPtr(handle), request.ToByteArray());
        }
        finally
        {
            handle.Free();
        }
    }

    private static T Parse<T>(Ffi.Dispatched d, MessageParser<T> parser, string what)
        where T : IMessage<T>
    {
        if (d.Status != 0)
        {
            throw Statuses.FromStatusBytes(d.Response, d.Status);
        }
        try
        {
            return parser.ParseFrom(d.Response ?? Array.Empty<byte>());
        }
        catch (InvalidProtocolBufferException e)
        {
            throw CodedError.Unhandled($"unmarshal {what}: {e.Message}");
        }
    }

    // --- invokers (type-erased bridges; the lone (TState) cast lives here) ----

    private static Any AnyOf(string typeUrl, byte[] payload) =>
        new() { TypeUrl = typeUrl, Value = ByteString.CopyFrom(payload) };

    private static Destinations DestinationsOf(
        Google.Protobuf.Collections.MapField<string, uint> seqs
    ) => new(new Dictionary<string, uint>(seqs));

    private static Invoker ApplierInvoker<TState>(Func<TState> factory, ApplierThunk<TState> thunk)
        where TState : class, IMessage =>
        (session, typeUrl, payload, aux) =>
        {
            thunk((TState)session.EnsureState(factory), AnyOf(typeUrl, payload));
            return new InvokerResult(null, Ffi.StatusOk);
        };

    private static Invoker CommandInvoker<TState>(Func<TState> factory, CommandThunk<TState> thunk)
        where TState : class, IMessage =>
        (session, typeUrl, payload, aux) =>
        {
            var cax = Abi.CommandContextAux.Parser.ParseFrom(aux);
            var cctx = new CommandContext(cax.NextSequence, cax.HadPriorEvents);
            var book = thunk(AnyOf(typeUrl, payload), (TState)session.EnsureState(factory), cctx);
            return book == null
                ? new InvokerResult(null, Ffi.StatusOkEmpty)
                : new InvokerResult(book.ToByteArray(), Ffi.StatusOk);
        };

    private static Invoker RejectionInvoker<TState>(
        Func<TState> factory,
        RejectionThunk<TState> thunk
    )
        where TState : class, IMessage =>
        (session, typeUrl, payload, aux) =>
        {
            var rax = Abi.RejectionAux.Parser.ParseFrom(aux);
            var n = Notification.Parser.ParseFrom(rax.Notification);
            var rej = RejectionNotification.Parser.ParseFrom(rax.Rejection);
            var cctx =
                rax.Cctx != null
                    ? new CommandContext(rax.Cctx.NextSequence, rax.Cctx.HadPriorEvents)
                    : new CommandContext(0, false);
            var resp = thunk(n, rej, (TState)session.EnsureState(factory), cctx);
            return resp == null
                ? new InvokerResult(null, Ffi.StatusOkEmpty)
                : new InvokerResult(resp.ToByteArray(), Ffi.StatusOk);
        };

    private static Invoker ProjectorEventInvoker<TState>(
        Func<TState> factory,
        ProjectorEventThunk<TState> thunk
    )
        where TState : class, IMessage =>
        (session, typeUrl, payload, aux) =>
        {
            thunk((TState)session.EnsureState(factory), AnyOf(typeUrl, payload));
            return new InvokerResult(null, Ffi.StatusOk);
        };

    private static Invoker ProjectorFinishInvoker<TState>(
        Func<TState> factory,
        ProjectorFinishThunk<TState> thunk
    )
        where TState : class, IMessage =>
        (session, typeUrl, payload, aux) =>
        {
            var book = EventBook.Parser.ParseFrom(payload);
            var proj = thunk((TState)session.EnsureState(factory), book);
            return new InvokerResult(proj.ToByteArray(), Ffi.StatusOk);
        };

    private static Invoker ProjectorUnknownInvoker(ProjectorUnknownThunk thunk) =>
        (session, typeUrl, payload, aux) =>
        {
            thunk(typeUrl);
            return new InvokerResult(null, Ffi.StatusOk);
        };

    private static Invoker SagaEventInvoker(SagaEventThunk thunk) =>
        (session, typeUrl, payload, aux) =>
        {
            var sax = Abi.SagaEventAux.Parser.ParseFrom(aux);
            var dests = DestinationsOf(sax.DestinationSequences);
            var emission = thunk(AnyOf(typeUrl, payload), dests, sax.SourceCover);
            var resp = new SagaResponse();
            resp.Commands.AddRange(emission.Commands);
            resp.Events.AddRange(emission.Events);
            return new InvokerResult(resp.ToByteArray(), Ffi.StatusOk);
        };

    private static Invoker SagaRejectionInvoker(SagaRejectionThunk thunk) =>
        (session, typeUrl, payload, aux) =>
        {
            var rax = Abi.RejectionAux.Parser.ParseFrom(aux);
            var n = Notification.Parser.ParseFrom(rax.Notification);
            var rej = RejectionNotification.Parser.ParseFrom(rax.Rejection);
            var resp = new SagaResponse();
            resp.Events.AddRange(thunk(n, rej));
            return new InvokerResult(resp.ToByteArray(), Ffi.StatusOk);
        };

    private static Invoker PmEventInvoker<TState>(Func<TState> factory, PmEventThunk<TState> thunk)
        where TState : class, IMessage =>
        (session, typeUrl, payload, aux) =>
        {
            var pax = Abi.PmEventAux.Parser.ParseFrom(aux);
            var dests = DestinationsOf(pax.DestinationSequences);
            var resp = thunk(AnyOf(typeUrl, payload), (TState)session.EnsureState(factory), dests);
            return new InvokerResult(resp.ToByteArray(), Ffi.StatusOk);
        };

    private static Invoker PmRejectionInvoker<TState>(
        Func<TState> factory,
        PmRejectionThunk<TState> thunk
    )
        where TState : class, IMessage =>
        (session, typeUrl, payload, aux) =>
        {
            var rax = Abi.RejectionAux.Parser.ParseFrom(aux);
            var n = Notification.Parser.ParseFrom(rax.Notification);
            var rej = RejectionNotification.Parser.ParseFrom(rax.Rejection);
            var r = thunk(n, rej, (TState)session.EnsureState(factory));
            var resp = new ProcessManagerHandleResponse();
            resp.ProcessEvents.AddRange(r.ProcessEvents);
            if (r.Escalation != null)
            {
                resp.Notification = r.Escalation;
            }
            return new InvokerResult(resp.ToByteArray(), Ffi.StatusOk);
        };
}
